java-security/src/main/java/com/sap/cloud/security/spring/adapter/SAPOfflineTokenServicesCloud.java

```diff
@@ -0,0 +1,59 @@
+package com.sap.cloud.security.spring.adapter;
+
+import com.sap.cloud.security.config.OAuth2ServiceConfiguration;
+import com.sap.cloud.security.token.Token;
+import com.sap.cloud.security.token.XsuaaToken;
+import com.sap.cloud.security.token.validation.CombiningValidator;
+import com.sap.cloud.security.token.validation.Validator;
+import com.sap.cloud.security.token.validation.validators.JwtValidatorBuilder;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.oauth2.common.OAuth2AccessToken;
+import org.springframework.security.oauth2.common.exceptions.InvalidTokenException;
+import org.springframework.security.oauth2.provider.AuthorizationRequest;
+import org.springframework.security.oauth2.provider.OAuth2Authentication;
+import org.springframework.security.oauth2.provider.token.ResourceServerTokenServices;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class SAPOfflineTokenServicesCloud implements ResourceServerTokenServices, InitializingBean {
+
+	private final Supplier<Validator<Token>> validatorSupplier;
+	private final OAuth2ServiceConfiguration serviceConfiguration;
+	private Validator<Token> tokenValidator;
+
+	public SAPOfflineTokenServicesCloud(OAuth2ServiceConfiguration serviceConfiguration) {
+		this(serviceConfiguration, () -> JwtValidatorBuilder.getInstance(serviceConfiguration).build());
+	}
+
+	SAPOfflineTokenServicesCloud(OAuth2ServiceConfiguration serviceConfiguration, Supplier<Validator<Token>> validatorSupplier) {
+		this.serviceConfiguration = serviceConfiguration;
+		this.validatorSupplier = validatorSupplier;
+	}
+
+	@Override
+	public OAuth2Authentication loadAuthentication(String accessToken)
```

Comment:
is accessToken always != null?

-> have tests, when token is invalid (IllegalArgumentException)-> an invalidTokenException should be raised

search:
https://github.wdf.sap.corp/CPSecurity/java-container-security/search?q=InvalidTokenException&unscoped_q=InvalidTokenException

Reply:
I agree we should be more defensive here and check it!